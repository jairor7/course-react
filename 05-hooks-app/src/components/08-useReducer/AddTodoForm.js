import React from "react";
import PropTypes from "prop-types";
import { useForm } from "../../hooks/useForm";

const AddTodoForm = ({ handleAddTodo }) => {
  const [{ description }, handleInputChange, reset] =
    useForm({
      description: "",
    });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length === 0) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      description,
      done: false,
    };
    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4>Agregar TODO</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          placeholder="New TODO"
          name="description"
          autoComplete="off"
          className="form-control"
          value={description}
          onChange={handleInputChange}
        />
        <div className="d-grid">
          <button
            type="submit"
            className="btn btn-outline-primary btn-block mt-3"
          >
            Agregar
          </button>
        </div>
      </form>
    </>
  );
};

AddTodoForm.propTypes = {
  handleAddTodo: PropTypes.func.isRequired,
};

export default AddTodoForm;
