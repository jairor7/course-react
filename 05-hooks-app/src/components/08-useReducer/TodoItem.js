import React from "react";
import PropTypes from "prop-types";

const TodoItem = ({
  todo: t,
  index,
  handleComplete,
  handleRemoveTodo,
}) => {
  return (
    <li key={t.id} className="list-group-item">
      <p
        className={`todo-task${t.done ? " complete" : ""}`}
        onClick={() => handleComplete(t.id)}
      >
        {index + 1}. {t.description}
      </p>
      <button
        className="btn btn-danger"
        onClick={() => handleRemoveTodo(t.id)}
      >
        Borrar
      </button>
    </li>
  );
};

TodoItem.propTypes = {
  todo: PropTypes.object.isRequired,
  // index: PropTypes.number.isRequired,
  // handleComplete: PropTypes.func.isRequired,
  // handleRemoveTodo: PropTypes.func.isRequired,
};

export default TodoItem;
