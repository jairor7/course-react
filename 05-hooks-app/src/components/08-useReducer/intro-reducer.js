const initialState = [
  {
    id: 1,
    todo: "Learn Hooks",
    done: false,
  },
  {
    id: 2,
    todo: "Learn React",
    done: false,
  },
];

const todoReducer = (state = initialState, action) => {
  switch (action?.type) {
    case "ADD_TODO":
      return [...state, action.payload];
    default:
      return state;
  }
};

let todos = todoReducer();

const newTodo = {
  id: 3,
  todo: "Learn Redux",
  done: false,
};

const addTodoAction = {
  type: "ADD_TODO",
  payload: newTodo,
};

todos = todoReducer([...todos], addTodoAction);

console.log(todos);
