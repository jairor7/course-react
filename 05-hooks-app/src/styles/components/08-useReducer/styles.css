body {
  padding: 50px;
}

p.todo-task {
  cursor: pointer;
  margin: 0;
}

.list-group-item {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.complete {
  text-decoration: line-through;
}
