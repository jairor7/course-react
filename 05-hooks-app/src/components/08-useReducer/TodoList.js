import React from "react";
import PropTypes from "prop-types";
import TodoItem from "./TodoItem";

const TodoList = ({
  todos,
  handleComplete,
  handleRemoveTodo,
}) => {
  return (
    <ul className="list-group list-group-flush">
      {todos.map((todo, index) => (
        <TodoItem
          key={todo.id}
          todo={todo}
          index={index}
          handleComplete={handleComplete}
          handleRemoveTodo={handleRemoveTodo}
        />
      ))}
    </ul>
  );
};

TodoList.propTypes = {
  todos: PropTypes.array.isRequired,
  handleComplete: PropTypes.func.isRequired,
  handleRemoveTodo: PropTypes.func.isRequired,
};

export default TodoList;
