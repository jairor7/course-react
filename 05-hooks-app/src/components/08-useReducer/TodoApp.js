import React, { useEffect, useReducer } from "react";
import { todoReducer } from "./todoReducer";
import "../../styles/components/08-useReducer/styles.css";
import TodoList from "./TodoList";
import AddTodoForm from "./AddTodoForm";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) ?? [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(
    todoReducer,
    null,
    init
  );

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const handleRemoveTodo = (id) => {
    const action = {
      type: "REMOVE_TODO",
      payload: id,
    };
    dispatch(action);
  };

  const handleComplete = (id) => {
    const action = {
      type: "IS_COMPLETE_TODO",
      payload: id,
    };

    dispatch(action);
  };

  const handleAddTodo = (newTodo) => {
    const action = {
      type: "ADD_TODO",
      payload: newTodo,
    };
    dispatch(action);
  };

  return (
    <div>
      <h1>Todo App ({todos.length})</h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleComplete={handleComplete}
            handleRemoveTodo={handleRemoveTodo}
          />
        </div>

        <div className="col-5">
          <AddTodoForm handleAddTodo={handleAddTodo} />
        </div>
      </div>
    </div>
  );
};
