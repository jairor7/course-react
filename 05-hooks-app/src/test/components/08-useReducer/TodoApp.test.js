import { TodoApp } from "../../../components/08-useReducer/TodoApp";
import { shallow, mount } from "enzyme";
import { demoInitialStateTodos } from "../../fixtures/demotodos";
import { act } from "@testing-library/react";

describe("Prueba del componente <TodoApp />", () => {
  Storage.prototype.setItem = jest.fn();
  const wrapper = shallow(<TodoApp />);
  test("Debe de mostrar correctamente el componente <TodoApp />", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de agregar un TODO", () => {
    const wrapper = mount(<TodoApp />);
    act(() => {
      wrapper.find("AddTodoForm").prop("handleAddTodo")(
        demoInitialStateTodos[0]
      );
      wrapper.find("AddTodoForm").prop("handleAddTodo")(
        demoInitialStateTodos[1]
      );
    });
    expect(wrapper.find("h1").text().trim()).toBe(
      "Todo App (2)"
    );
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    expect(localStorage.setItem).toHaveBeenCalledWith(
      "todos",
      '[{"id":1,"description":"Hacer la compra","done":false},{"id":2,"description":"Lavar la loza","done":false}]'
    );
  });
  test("Debe de eleminar un TODO", () => {
    wrapper.find("AddTodoForm").prop("handleAddTodo")(
      demoInitialStateTodos[0]
    );
    expect(wrapper.find("h1").text().trim()).toBe(
      "Todo App (1)"
    );
    wrapper.find("TodoList").prop("handleRemoveTodo")(
      demoInitialStateTodos[0].id
    );
    expect(wrapper.find("h1").text().trim()).toBe(
      "Todo App (0)"
    );
  });
});
