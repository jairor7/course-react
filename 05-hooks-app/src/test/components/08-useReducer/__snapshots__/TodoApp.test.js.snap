// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba del componente <TodoApp /> Debe de mostrar correctamente el componente <TodoApp /> 1`] = `
<div>
  <h1>
    Todo App (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleComplete={[Function]}
        handleRemoveTodo={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <AddTodoForm
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
